Add Compound USDC Borrow APR as Price Identifier

evanmays opened the pull request.

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+Cost: Calculating an aggregated statistic of compound USDC borrow rates on a confirmed range of blocks on the Ethereum blockchain is easy. All of the needed data can be computed from event logs on any Ethereum full node. An archive node allows direct querying of this data via contract methods. Additionally, this data can be accessed through querying the Compound Subgraph on the Graph Protocol or  ...
+
+Opportunity: Compound USDC Borrow APR futures can help borrowers and traders hedge against interest rate volatility allowing them to "lock-in" future costs from Compound loans. This will enable taking multi-collateral backed loans out on Compound for new use cases. Other users may wish to speculate on the demand for borrowing USDC on Compound or hedge their deposit interest rate volatility. Provi ...
+
```
Also note the "Price feed Implementation" section in the guide. A PR for the price feed is required.

Tommy1231232 replied:
@evanmays were you able to make any progess over the past week? Would be great to get this up for a vote this week :)

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+## Technical Specification
+The definition of this identifier should be:
+
+* Identifier name: COMPUSDCAPR-1M
```
If your identifier name is greater than 8 characters, the price ID should be written as [BASE/QUOTE]. See 'Technical Specification' section in the guide for more details

Tommy1231232 replied:
What does the 1M mean?

evanmays replied:
1M is supposed to be 1 month. I'll add that in writing

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+## Headers
+- UMIP <#>
+- UMIP title: Add COMPUSDCAPR-1M as a price identifier
```
Update name format per below comment

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+* Base Currency: Compound Borrowing USDC Interest Annual Percentage Rate
+* Quote Currency: USD
+* Sources: Derivable from data on any Ethereum full node or data set of Compound Protocol data
+* Processing: Use the borrow rate per block to calculate the average APR (geometric mean) over a rolling 30 day period (See implementation section)
```
Per the tech specs in the guide, can you add fields for:

-Intended collateral currency
-Collateral decimals

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+* Output processing: Convert APR percentage to USD 1-to-1. (Example: 7% is $7USD and 3% is $3USD)
+
+
+## Rationale
```
This was a really well written section! Awesome job

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+
+We mitigate the effects of numerical instability by rounding to the nearest two decimal places. Different algorithms for calculating the geometric mean result in tiny differences in the result. Rounding to 2 decimal places hides small differences in geometric mean calculations. For example, if person A calculates the price request result as $7.53453USD and person B calculates the price request re ...
+
+## Implementation
```
See the "Implementation" section in the price ID guide for questions to address in this section. We need to add:

-Step by step directions for someone to calculate the price

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+* Base Currency: Compound Borrowing USDC Interest Annual Percentage Rate
+* Quote Currency: USD
+* Sources: Derivable from data on any Ethereum full node or data set of Compound Protocol data
+* Processing: Use the borrow rate per block to calculate the average APR (geometric mean) over a rolling 30 day period (See implementation section)
```
I believe you meant this to be "input processing". If so, this relates to denoting if and why UMA tokenholders should change the calculation method for the price identifier. This is usually denoted with ""Human intervention in extreme circumstances where the result differs from broad market consensus"

evanmays replied:
Gotcha. I wasn't sure what exactly the different types of processing meant before. I understand now. Also I see the definitions in the new guide

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+* Rounding: Round to nearest 2 decimal places (third decimal place digit >= 5 rounds up and < 5 rounds down)
+* Pricing Interval: Per block
+* Dispute timestamp rounding: Round down (previous timestamp)
+* Output processing: Convert APR percentage to USD 1-to-1. (Example: 7% is $7USD and 3% is $3USD)
```
Result processing specifies what the DVM does once it obtains all the votes from UMA token holders (e.g., take the median or mode of all votes submitted). Specify if you want the median or mode and then have the DVM perform the conversion from % to USD.

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+## Headers
```
FYI there is a new format for creating a UMIP for price identifiers. Since you started this before the new template was released I'll leave it up to you if you want to make it look similar to the format of the guide. I just want to make sure responses to questions in the guide are stated somewhere in here. 

https://github.com/UMAprotocol/UMIPs/blob/master/price-identifier-guide.md

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+We mitigate the effects of numerical instability by rounding to the nearest two decimal places. Different algorithms for calculating the geometric mean result in tiny differences in the result. Rounding to 2 decimal places hides small differences in geometric mean calculations. For example, if person A calculates the price request result as $7.53453USD and person B calculates the price request re ...
+
+## Implementation
+TODO: Implement getBorrowRatePerBlock() function and test implementation works in python and is numerically stable compared to alternative implementations in different programming languages and computing environments
```
I see is says TODO, let me know when done so we can get this approved! Really great product :)

evanmays commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,77 @@
+
+Cost: Calculating an aggregated statistic of compound USDC borrow rates on a confirmed range of blocks on the Ethereum blockchain is easy. All of the needed data can be computed from event logs on any Ethereum full node. An archive node allows direct querying of this data via contract methods. Additionally, this data can be accessed through querying the Compound Subgraph on the Graph Protocol or  ...
+
+Opportunity: Compound USDC Borrow APR futures can help borrowers and traders hedge against interest rate volatility allowing them to "lock-in" future costs from Compound loans. This will enable taking multi-collateral backed loans out on Compound for new use cases. Other users may wish to speculate on the demand for borrowing USDC on Compound or hedge their deposit interest rate volatility. Provi ...
```
Yes, will add example! Planning on stable coin as collateral (probably USDC). This should make the economic math easier for people who want to open short positions

evanmays commented:
Great comments! I'll address everything and push the commits

evanmays commented:
I added the implementation and addressed the other comments. I don't think we need a liquidator bot price feed since this price identifier will only be used at expiration. Similar to how GASETH-FEB21 (UMIP-25) only uses the historical data for the expiration price.

I'll need to work on another UMIP for the two hour TWAP which tracks the synthetic token similar to GASETH-FEB21

rubberbandedman commented:
why would the price feed implementation for the liquidator bot not be useful here...? confused, don't people in the uma ecosystem need to run liquidator bots upon synth launch?

evanmays commented:
@rubberbandedman The intention is that this price identifier won't be used for a trading synthetic. This price identifier is only used on the expiration date of a synthetic. By the time this price identifier will be used (at expiration), liquidations won't be possible anymore.

There's more about this in https://github.com/UMAprotocol/UMIPs/pull/150 as well as UMIP-22 (uGAS) which has a similar mechanism.

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,130 @@
+## Technical Specification
+The definition of this identifier should be:
+
+* Identifier name: [COMPUSDCAPR-30DAY/USD]
```
Suggest changing the name to [COMPUSD-APR-1M]/USD

Removing the C in USDC makes it more generalized and easier to read with the hyphen before APR.

evanmays replied:
Agreed on the APR. Would like to hear more of your thoughts on the USDC though.

The C is intended to make it clear that this is the APR to borrow USDC on Compound as opposed to DAI or some other stable coin. I think maybe it would be confusing to say USD since you can't borrow USD on Compound. Open to disagreeal though if I'm being too pedantic

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,130 @@
+Raw data is also available for download at the [Tendies Exchange public endpoint](https://cache.tendies.exchange/borrow_rate_per_block.json). This endpoint is updated with every new block, but delayed by 20 block confirmations. This endpoint is free to use. To generate a similar file, governors can use and/or modify this [open source indexer script](https://github.com/evanmays/tendies-exchange/tr ...
+
+All of these endpoints give the borrow rate per block for historical blocks, however converting this data to APR (and aggregation) is still required. (Read the implementation section)
+
```
Can you add a section on "Price Feed Implementation"

See guide: https://github.com/UMAprotocol/UMIPs/blob/master/price-identifier-guide.md

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,130 @@
+
+## Implementation
+
+This implementation works with the dataset from [Tendies Exchange public endpoint](https://cache.tendies.exchange/borrow_rate_per_block.json). To use it with other datasets, modify the `getBorrowRatePerBlock()` function. This implementation is updated as an [open source price feed script](https://github.com/evanmays/tendies-exchange/tree/master/indexer).
```
@mrice32 can you confirm there needs to be a new price feed comprised of the below?

Tommy1231232 commented:
> why would the price feed implementation for the liquidator bot not be useful here...? confused, don't people in the uma ecosystem need to run liquidator bots upon synth launch?



> @rubberbandedman The intention is that this price identifier won't be used as a peg target for a synthetic. This price identifier is only used on the expiration date of a synthetic. By the time this price identifier will be used (at expiration), liquidations won't be possible anymore.
> 
> There's more about this in #150 as well as UMIP-22 (uGAS) which has a similar mechanism.

Can we still add the Price Feed section and detail out the above?

Tommy1231232 commented:
> price identifier will only be used at expiration. Similar to how GASETH-FEB21 (UMIP-25) only uses the historical data for the expiration price.

@smb2796 can you confirm this? Wouldn't bots still need to constantly query the price to determine what position to liquidate?

evanmays commented:
@Tommy1231232 I put more details about the price feed in new Price Feed Implementation section. It should answer all your questions but let me know if not.

Also added hyphen, but left as USDC for now

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,134 @@
+
+The source of truth for this data is the Compound USDC cToken's (cUSDC) `borrowRatePerBlock()` method. As of the writing of this UMIP, the agreed upon cUSDC smart contract address is `0x39aa39c021dfbae8fac545936693ac917d5e7563`. This price identifier aggregates the value returned by `borrowRatePerBlock()` over every block from the 30 days ending at the DVM timestamp.
+
+It is recommended to gather the raw data from an Ethereum archive node. Alternatively, this data is indexed in the [Graph Protocol](https://thegraph.com/explorer/subgraph/graphprotocol/compound-v2). As of writing this UMIP, Graph Protocol is free. However, they have plans to start charging for access in the future. In the future, querying 30 days worth of blocks may cost up to $20 USD. This UMIP' ...
```
Unless this is used with a different contract type that has a different mechanism for ensuring overcollateralization, this would still need to be used with liquidator and dispute bots.

evanmays replied:
Overcollateralization is determined with a different price identifier: The two-hour TWAP specified in https://github.com/UMAprotocol/UMIPs/pull/150

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,134 @@
+## Price Feed Implementation
+
+
+The intention is that this price identifier won't be used as a peg target for a synthetic. This price identifier is only used on the expiration date of a synthetic. By the time this price identifier will be used (at expiration), liquidations won't be possible anymore. Therefore, a price feed isn't required for liquidator bots to track this price identifier. All liquidations will use a two-hour TW ...
```
What was the reasoning for splitting this up between two price identifiers/UMIPs? I think a better way to do this would be to make this one price identifier and have this calculation methodology be used after the cutoff time specified in #150 and use the price calculation method in #150 before the expiration timestamp. This is how it is accomplished with GASETH, so I suggest merging this into one pid UMIP.

evanmays replied:
That is currently the plan. https://github.com/UMAprotocol/UMIPs/pull/150 relies on this price identifier. And this price identifier will only be used for https://github.com/UMAprotocol/UMIPs/pull/150. I think the uGAS GASETH UMIPs were separate to have each UMIP be more focused. Happy to merge the two if people prefer

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,134 @@
+* Sources: Derivable from data on any Ethereum full node or data set of Compound Protocol data
+* Rounding: Round to nearest 2 decimal places (third decimal place digit >= 5 rounds up and < 5 rounds down)
+* Input processing: If the cUSDC smart contract has an unexpected failure or hack, human intervention is required. For UMA tokenholders without access to archive nodes, in the case that [Tendies Exchange public endpoint](https://cache.tendies.exchange/borrow_rate_per_block.json) is down, tokenholders should use [Graph Protocol](https://thegraph.com/explorer/subgraph/graphprotocol/compound-v2).
+* Result Processing: Take the median of all votes submitted.
```
I don't think this should be included since the price identifier cant change the behavior of the dvm. The DVM will really take the mode of all votes submitted.

evanmays replied:
Fixed

evanmays commented:
Going to combine #150 into this PR

mhairimca commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+Post Cutoff Cost: Calculating an aggregated statistic of compound USDC borrow rates on a confirmed range of blocks on the Ethereum blockchain is easy. All of the needed data can be computed from event logs on any Ethereum full node. An archive node allows direct querying of this data via contract methods. Additionally, this data can be accessed through querying the Compound Subgraph on the Graph  ...
+
+Opportunity: The usage of a two-hour twap for liquidations, then a aggregated price at expiration is similar to UMIP-22. This setup allows us to create futurues. Compound USDC Borrow APR futures can help borrowers and traders hedge against interest rate volatility allowing them to "lock-in" future costs from Compound loans. This will enable taking multi-collateral backed loans out on Compound for ...
```
typo in "futures"

evanmays replied:
good catch

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+**5. Collateral Decimals** - 6
+
+**6a. Rounding Pre Cutoff** - 6 decimal places. (decimal place >= 5 rounds up)
```
```suggestion
**6a. Rounding Pre Cutoff** - 6 decimal places. (7th decimal place >= 5 rounds up)
```

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+**6a. Rounding Pre Cutoff** - 6 decimal places. (decimal place >= 5 rounds up)
+
+**6b. Rounding Post Cutoff** - Round to nearest 2 decimal places (third decimal place digit >= 5 rounds up and < 5 rounds down)
```
Why would the rounding change Pre to Post cutoff? The price should in the same order of magnitude, so shouldnt the rounding stay the same?

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+## Technical Specification
+
+### February Price Identifier
```
```suggestion
### COMPUSDC-APR-TWAP-OR-30DAY-FEB28/USD
```

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+**9. Dispute timestamp rounding** Round down (previous timestamp)
+
+### March Price Identifier
```
```suggestion
### COMPUSDC-APR-TWAP-OR-30DAY-MAR28/USD
```

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+**6a. Rounding Pre Cutoff** - 6 decimal places. (decimal place >= 5 rounds up)
+
+**6b. Rounding Post Cutoff** - Round to nearest 2 decimal places (third decimal place digit >= 5 rounds up and < 5 rounds down)
```
Same comment as above about rounding consistency

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+### Pre-cutoff
+
+Due to the lack of ambiguity on calculating the TWAP, UMA token holders should all converge on the same price. Therefore, we don't require rounding pre cutoff.
```
I'd still recommend keeping this consistent between pricing times. The possible voting disagreements introduced by numerical instability in the Post-cutoff price could also be introduced by just general lack of awareness/confusion from a rounding switch.

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+3. A single Uniswap/Balancer price is defined for each timestamp as the price that the ETH/uGAS pool returns at the end of the latest block whose timestamp is <= the timestamp that is queried for.
+4. The TWAP is an average of the prices for each timestamp between the start and end timestamps. Each price in this average will get an equal weight.
+5. As the token price will already implicitly be tracking the GASETH-1M-1M price, it should be left as returned without any scaling transformation.
+6. The final price should be returned with the synthetic token as the denominator of the price pair and should be submitted with 18 decimals.
```
```suggestion
6. The final price should be returned with the synthetic token as the denominator of the price pair and should be rounded to 6 decimals.
```

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+2. The start TWAP timestamp is defined by the end TWAP timestamp minus the TWAP period (2 hours in this case).
+3. A single Uniswap/Balancer price is defined for each timestamp as the price that the ETH/uGAS pool returns at the end of the latest block whose timestamp is <= the timestamp that is queried for.
+4. The TWAP is an average of the prices for each timestamp between the start and end timestamps. Each price in this average will get an equal weight.
+5. As the token price will already implicitly be tracking the GASETH-1M-1M price, it should be left as returned without any scaling transformation.
```
```suggestion
5. As the token price will already implicitly be tracking the COMPUSDC-APR-TWAP-OR-30DAY-FEB28/USD or COMPUSDC-APR-TWAP-OR-30DAY-MAR28/USD price, it should be left as returned without any scaling transformation.
```

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+1. The end TWAP timestamp equals the price request timestamp.
+2. The start TWAP timestamp is defined by the end TWAP timestamp minus the TWAP period (2 hours in this case).
+3. A single Uniswap/Balancer price is defined for each timestamp as the price that the ETH/uGAS pool returns at the end of the latest block whose timestamp is <= the timestamp that is queried for.
```
```suggestion
3. A single Uniswap/Balancer price is defined for each timestamp as the price that the synthetic COMPUSDC-APR/USDC pool returns at the end of the latest block whose timestamp is <= the timestamp that is queried for.
```

evanmays replied:
going to do USDC/CAR pool to keep consistent with the synthetic asset name as CAR

evanmays commented:
Thanks for the feedback @smb2796
I'll change everything to 2 decimal places of rounding

evanmays commented:
I'm open to changing everything to rounding to 6 decimal places (which is consistent with the collateral currency decimal places) if folks think numerical instability won't crop up. I could imagine the argument that the voting community can come to consensus on the exact price before submitting their votes.

nicholaspai commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+## Summary
+
+The DVM should support price requests for the [COMPUSDC-APR-TWAP-OR-30DAY-FEB28/USD] and [COMPUSDC-APR-TWAP-OR-30DAY-MAR28/USD] price indices.
```
You don't actually have to define the expiration timestamps in the UMIP for a price identifier. Assuming the CAR identifier gets whitelisted, then you (or anyone) are free to create as many financial contracts as you want that use the CAR identifier and expire to whatever timestamp you want.

evanmays replied:
I guess that is possible. It seems like that approach may confuse tokenholders though. We’d be putting information needed for voting (the cutoff timestamp) outside of the UMIP. Voters would need to check exact expiration timestamps on chain as opposed to just reading the UMIP.

nicholaspai replied:
> Voters would need to check exact expiration timestamps on chain as opposed to just reading the UMIP.

This is how all price requests work currently. Voters look at a price request's "identifier" and the timestamp at which it was created, in order to resolve a price. The expiration timestamp for a particular contract is just an on-chain enforced time at which the expiry price request for CAR will be sent.

smb2796 replied:
The timestamps are included because this price identifier needs to switch its calculation methodology at expiry. Without ancillary data, I believe this is needed.

evanmays replied:
Timestamps can be removed from the identifier. But we will be left with an identifier that behaves differently depending on which EMP (or other financial contract) calls it. 

By removing the timestamp, voters will have to look at the price request's "identifier", the timestamp of the DVM request, and the code of the smart contract that called the DVM request.

By keeping the timestamp, voters don't need to know anything about the code of the smart contract that called the DVM request

chrismaree replied:
I think the timestamps should be kept. in this way it follows the same pattern as uGAS which defines a once off usage identifier that transforms at settlement: https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-25.md#implementation

In the future, when we have ancillary data & the ability to use `FinancialProductLibrary` to `transformPriceIdentifier` we could do this with one identifier for any variant of this kind of product. but for now this is not yet on mainnet nor is it ready for prime time :)

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -1,56 +1,135 @@
 ## Headers
 - UMIP <#>
-- UMIP title: Add [COMPUSDC-APR-30DAY/USD] as a price identifier
+- UMIP title: Price identifiers for CAR: [COMPUSDC-APR-TWAP-OR-30DAY-FEB28/USD] [COMPUSDC-APR-TWAP-OR-30DAY-MAR28/USD]
```
No need to specify the synthetic asset you plan to use that references this price ID. Keep it as "Add [COMPUSDC -30day/USD] as a price identifier" (note, no need for dates).

smb2796 replied:
I do think the dates are needed and should be included since these pids are specified for certain expiry timestamps

evanmays replied:
I updated the title to not mention CAR. Left dates in for the reason @smb2796 described

chrismaree commented on UMIPs/umip-draft_compound_borrow_apr.md:
````diff
@@ -0,0 +1,229 @@
+The CAR token tied to this price identifier will expire on the cutoff timestamp `1614470400`.
+```
+CUTOFF = 1614470400 # Feb 28, 2021 00:00:00 UTC
+if dvm_timestamp >= CUTOFF:
````
When we have the `FinancialProductLibraries` this logic will become codified within the EMP itself, rather than the UMIP https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/financial-templates/common/financial-product-libraries/PreExpirationIdentifierTransformationFinancialProductLibrary.sol

will be great.

evanmays replied:
I can't wait!!

chrismaree commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+
+Pre Cutoff Cost: Calculating the two-hour TWAP is an easy calculation. The reference UMA liquidator bot implementation already has a two-hour TWAP Uniswap/Balancer implementation.
+
+Post Cutoff Cost: Calculating an aggregated statistic of compound USDC borrow rates on a confirmed range of blocks on the Ethereum blockchain is easy. All of the needed data can be computed from event logs on any Ethereum full node. An archive node allows direct querying of this data via contract methods. Additionally, this data can be accessed through querying the Compound Subgraph on the Graph  ...
```
It's convenient that we don't need aggregated rates in the bots post-settlement & the bots will work by default with the synth as it is using a self preferential TWAP for liquidations.

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+## Headers
+- UMIP <#>
+- UMIP title: Add price identifiers [COMPUSDC-APR-TWAP-OR-30DAY-FEB28/USD] and [COMPUSDC-APR-TWAP-OR-30DAY-MAR28/USD]
```
Change USD to USDC in the name

evanmays replied:
The USD is the quote currency, I think we want to keep it that way unless there's a good reason for using USDC as quote currency.

evanmays replied:
As a different example (not in the UMIP at all). Suppose we wan't to track the ETH borrow rates on Compound. We would name that price identifier [COMPETH-APR-TWAP-OR-30DAY-FEB28/USD]

evanmays replied:
Nevermind. I see why we need this now. The bot's won't be tracking CAR/USD. They'll be tracking CAR/USDC. I'm changing as suggested

Tommy1231232 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,229 @@
+if dvm_timestamp >= CUTOFF:
+  Resolve price to the 30 day total annualized interest rate on borrowing USDC from Compound. (See implementation)
+else:
+  Resolve price to the 2-hour Time-Weighted Average Price for the Uniswap or Balancer price of the CAR token quoted in USD. CAR token address in technical specification.
```
For the TWAP lets specify to only use Uniswap

evanmays replied:
Sure

smb2796 commented on UMIPs/umip-draft_compound_borrow_apr.md:
```diff
@@ -0,0 +1,227 @@
+## Headers
+- UMIP <#>
```
```suggestion
- UMIP-38
```

smb2796 commented:
Hey Evan. Thanks for all of the adjustments. Can you change the file name to umip-38 so that this can be merged in?

evanmays commented:
Changed title, renamed file, and updated status to last call

smb2796 reviewed (approved):
Looks good. Going to approve and merge in as "Last Call".
